modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskProcessor.java

```diff
@@ -1551,4 +1597,65 @@ else if (msg instanceof GridTaskSessionRequest)
             ? ", task name: " + task.getSession().getTaskName()
             : "";
     }
+
+    /**
+     * Subscription to update the status of tasks.
+     *
+     * <p>NOTE: {@link ComputeGridMonitor#processStatusSnapshots} will be called only on subscription,
+     * then only {@link ComputeGridMonitor#processStatusDiff} will be called.
+     *
+     * @param monitor Task status update monitor.
+     * @throws IllegalStateException If the node is stopped.
+     */
+    public void listenStatusUpdates(ComputeGridMonitor monitor) {
+        lock.writeLock();
+
+        try {
+            if (stopping)
+                throw new IllegalStateException("Failed to add monitor due to grid shutdown: " + monitor);
+
+            taskStatusMonitors.add(monitor);
+
+            monitor.processStatusSnapshots(taskStatusSnapshots.values());
+        }
+        finally {
+            lock.writeUnlock();
+        }
+    }
+
+    /**
+     * Unsubscribe to update the status of tasks.
+     *
+     * @param monitor Task status update monitor.
+     */
+    public void stopListenStatusUpdates(ComputeGridMonitor monitor) {
+        lock.writeLock();
+
+        try {
+            taskStatusMonitors.remove(monitor);
+        }
+        finally {
+            lock.writeUnlock();
+        }
+    }
+
+    /**
+     * Notify {@link #taskStatusMonitors} about task status changes.
+     *
+     * @param diff Task status diff.
+     * @param beforeNotify Function to be executed before notifying {@link #taskStatusMonitors}.
+     */
+    private void notifyTaskStatusMonitors(ComputeTaskStatusDiff diff, Runnable beforeNotify) {
+        lock.readLock();
+
+        try {
+            beforeNotify.run();
+
+            for (ComputeGridMonitor monitor : taskStatusMonitors)
+                monitor.processStatusDiff(diff);
```

Comment:
Can these guys throw exceptions? Maybe they can, we don't control it. Please surround every iteration with try/catch